File: frontend/src/entities/resources.ts
import * as serverV1Types from "../../lib/protobuf/server/v1/server_pb"

export default class Resources {
  time: number
  gold: number

  constructor(resources: Partial<serverV1Types.Resources>) {
    this.time = resources.time ?? 0
    this.gold = resources.gold ?? 0
  }

  multiply(multiplier: number): Resources {
    return new Resources({
      time: this.time * multiplier,
      gold: this.gold * multiplier,
    })
  }
}

File: cmd/http/model/resources.go
package model

import serverv1 "war-of-faith/pkg/protobuf/server/v1"

type Resources struct {
	Time uint32
	Gold uint32
}

func (r Resources) ToProtobuf() *serverv1.Resources {
	return &serverv1.Resources{
		Time: r.Time,
		Gold: r.Gold,
	}
}

func (r Resources) Multiply(multiplier uint32) Resources {
	return Resources{
		Time: r.Time * multiplier,
		Gold: r.Gold * multiplier,
	}
}
